### PR TITLE
Adapting generators to opmath enabled/disabled

### PR DESCRIPTION
**Pull request by PietropaoloFrisoni**
**Context:** After enabling the new operator arithmetic by default, we want the generators in the source code to return a `LinearCombination` instance or a `Hamiltonian` instance wherever possible.

**Description of the Change:** The generators touched in this PR are modified so that they return `qml.Hamiltonian` instead of `Sum`, `Prod`, or `Sprod` instances. When opmath is enabled, `qml.Hamiltonian` points to `pennylane.ops.op_math.linear_combination.LinearCombination`, and when it is disabled, it points to `pennylane.ops.qubit.hamiltonian.Hamiltonian`. This ensures that the appropriate instance is used consistently.
Note that the generators unchanged in this PR are modified (wherever possible) in #5410 , #5411 , #5412 (including the changelog entry).

**Benefits:** A more coherent choice depending on whether opmath is enabled or disabled. 

**Possible Drawbacks:** None that I can think of, except that old opmath would be deprecated in the future. Therefore, some precautions that have been taken here (to ensure that tests associated with generators work even with opmath disabled) might become useless.

**Related GitHub Issues:** None.

[sc-57982]

### CODE REVIEW

File: pennylane/ops/op_math/pow.py
Comment (PietropaoloFrisoni) [inline pennylane/ops/op_math/pow.py:1]: This `qml.simplify` is due to the following behavior:

```
>>> #self.base.generator() = -0.5 * X(0)
>>> #self.z = 2.5
>>> qml.Hamiltonian([2.5], [-0.5 * qml.X(0)]), qml.simplify(qml.Hamiltonian([2.5], [-0.5 * qml.X(0)]))
(2.5 * -0.5 * X(0), -1.25 * X(0))
```
If we call `qml.generator` as follows (without the `qml.simplify` here):
```
>>> z = 2.5
>>> base = qml.RX(2.34, wires=0)
>>> op = Pow(base, z)
>>> qml.generator(op, format="prefactor")
(-0.5 * X(0), 2.5)
```
That is, the `qml.generator(op, format="prefactor")` function requires the function to be simplified not to have bugs.

File: tests/ops/op_math/test_adjoint.py
Comment (PietropaoloFrisoni) [inline tests/ops/op_math/test_adjoint.py:1]: This is just to remark on the difference since `test_generator_legacy` would work even with opmath enabled (but `test_generator` does not work with opmath disabled) after the last modification to the source code. Probably it is best to eliminate the second one and remove the `@pytest.mark.usefixtures("use_legacy_opmath")` from the former

File: pennylane/ops/op_math/pow.py
Comment (PietropaoloFrisoni) [inline pennylane/ops/op_math/pow.py:1]: As for `adjoint.py`, here this extra line is only necessary when 
- opmath is **disabled**. That is, to avoid situations like:

```
>>> qml.Hamiltonian([-1.0], [op.generator()])
(-1.0) [Hamiltonian0]
```

- when opmath is **enabled**, but we don't simplify:


```
>>> qml.Hamiltonian([-1.0], [op.generator()])
-1.0 * 0.5 * X(0)
```

File: pennylane/ops/op_math/adjoint.py [hunk 1]
```diff
@@ -401,7 +401,8 @@ def has_generator(self):
         return self.base.has_generator
 
     def generator(self):
-        return qml.s_prod(-1.0, self.base.generator())
+        base_gen = self.base.generator()
+        return qml.Hamiltonian([-1.0 * base_gen.coeffs[0]], base_gen.ops)
```
Comment (lillian542) [inline pennylane/ops/op_math/adjoint.py:405]: I can see this used to be `-1 * self.base.generator()`, and then the `s_prod` above. What has changed so the we need to separate out the `base_gen.coeffs` now?

Comment (lillian542) [reply pennylane/ops/op_math/adjoint.py:405]: And what happens here if the `base_gen` is a `Sum`, for example? By unpacking `base_gen.ops`, but still only having a single coefficient, the length of coifs and base_gen won't match, and we'll get an error.

File: pennylane/ops/op_math/pow.py
Comment (lillian542) [reply pennylane/ops/op_math/pow.py:1]: Ah, I see - but this is broken if your base Hamiltonian has more than one term. Better to end up with something that could be simplified further than something that raises an error. Possibly best to just restore this to `-1 * base_gen`. Then it will return a `Hamiltonian` with legacy opmath and an `SProd` with new opmath. 

It doesn't follow the "things that are Hamiltonians with legacy functionality are LinearCombinations now" convention, but we're following that a lot more loosely than "if new_opmath is enabled, you should get an arithmetic operator, and if old opmath is enabled, a legacy operator".

File: pennylane/ops/op_math/pow.py [hunk 2]
```diff
@@ -329,7 +329,8 @@ def generator(self):
 
         See also :func:`~.generator`
         """
-        return qml.s_prod(self.z, self.base.generator(), lazy=False)
+        base_gen = self.base.generator()
+        return qml.Hamiltonian([self.z * base_gen.coeffs[0]], base_gen.ops)
```
Comment (lillian542) [inline pennylane/ops/op_math/pow.py:333]: same issue

File: pennylane/ops/qubit/parametric_ops_multi_qubit.py [hunk 3]
```diff
@@ -1236,12 +1239,12 @@ class IsingXY(Operation):
...
-            qml.sum(
-                qml.prod(PauliX(wires=self.wires[0]), PauliX(wires=self.wires[1])),
-                qml.prod(PauliY(wires=self.wires[0]), PauliY(wires=self.wires[1])),
-            ),
+        return qml.Hamiltonian(
+            [0.25],
+            [
+                PauliX(wires=self.wires[0]) @ PauliX(wires=self.wires[1])
+                + PauliY(wires=self.wires[0]) @ PauliY(wires=self.wires[1])
+            ],
```
Comment (lillian542) [inline pennylane/ops/qubit/parametric_ops_multi_qubit.py:1247]: Not sure why this one is an issue and everything else is fine, but this seems to break `parameter_frequencies` (and thereby adjoint differentiation) for `IsingXY`. 

To make sure we preserve functionality, let's add the `use_legacy_and_new_opmath` fixture to the parameter frequency test as well (getting parameter frequencies is one of the main uses for the generators). And then debug why this particular operator is failing 😅

File: pennylane/ops/op_math/adjoint.py
Comment (PietropaoloFrisoni) [inline pennylane/ops/op_math/adjoint.py:1]: The reason why we are not simply returning `qml.Hamiltonian([-1.0], [base.generator()])` is that, with opmath **disabled**, we have the following behavior that causes problems in many tests:
```
>>> qml.Hamiltonian([-1.0], [base.generator()])
(-1.0) [Hamiltonian0]
```

File: pennylane/ops/op_math/pow.py
Comment (PietropaoloFrisoni) [inline pennylane/ops/op_math/pow.py:1]: The reason why we are not simply returning `qml.Hamiltonian([self.z], [self.base.generator()])` is the same explained for the file `pennylane/ops/op_math/adjoint.py` (see comment above)

Comment (Qottmann) [reply pennylane/ops/op_math/pow.py:1]: Note that for LinearCombination class a scalar multiplication is eager:
```
>>> 0.5 * qml.ops.LinearCombination([0.5], [X(0)])
0.25 * X(0)
```

Doesnt that simplify the problem?

**Review by Qottmann — CHANGES_REQUESTED** [review]
I think we can have best of both worlds with the suggestion. Else looks great 👍 Happy to approve with that fixed :)

File: pennylane/ops/op_math/pow.py
Comment (lillian542) [reply pennylane/ops/op_math/pow.py:1]: Not all generators are `LinearCombinations`, but definitely makes it more of an edge case. What is missing should be handled elsewhere, not here. It shouldn't be a problem that things are `SProd` of `Sum` for example (IMO).

File: tests/ops/op_math/test_adjoint.py [hunk 4]
```diff
@@ -489,12 +489,20 @@ def test_has_generator_false(self):
 
         assert op.has_generator is False
 
+    @pytest.mark.usefixtures("use_legacy_opmath")
+    def test_generator_legacy(self):
+        """Assert that the generator of an Adjoint is -1.0 times the base generator with opmath disabled."""
```
Comment (lillian542) [inline tests/ops/op_math/test_adjoint.py:494]: Doesn't this version work for both now?

**Review by lillian542 — APPROVED** [review]
I think one pair of `legacy` and `new` tests can be combined now, but no blocking comments. Looks good to me, glad this is resolved! 🚀

File: pennylane/ops/op_math/pow.py
Comment (Qottmann) [reply pennylane/ops/op_math/pow.py:1]: yep sounds good 👍

**Review by Qottmann — APPROVED** [review]
nice 👌
